adapters/ix/ix.go

```diff
@@ -1,267 +1,189 @@
-			thisImp.Banner.H = &format.H
-			indexReq.Imp = []openrtb.Imp{thisImp}
-			// Index spec says "adunit path representing ad server inventory" but we don't have this
-			// ext is DFP div ID and KV pairs if avail
-			//indexReq.Imp[i].Ext = json.RawMessage("{}")
-
-			// Any objects pointed to by indexReq *must not be mutated*, or we will get race conditions.
-			siteCopy := *indexReq.Site
-			siteCopy.Publisher = &openrtb.Publisher{ID: params.SiteID}
-			indexReq.Site = &siteCopy
+	// Multi-size banner imps are split into single-size requests.
+	// The first size imp requests are added to the first slice.
+	// Additional size requests are added to the second slice and are merged with the first at the end.
+	// Preallocate the max possible size to avoid reallocating arrays.
+	requests := make([]*adapters.RequestData, 0, maxRequests)
+	multiSizeRequests := make([]*adapters.RequestData, 0, maxRequests-nImp)
+	errs := make([]error, 0, 1)
 
-			// spec also asks for publisher id if set
-			// ext object on request for prefetch
-			j, _ := json.Marshal(indexReq)
+	headers := http.Header{
+		"Content-Type": {"application/json;charset=utf-8"},
+		"Accept":       {"application/json"}}
 
-			request := callOneObject{requestJSON: *bytes.NewBuffer(j), width: format.W, height: format.H}
-
-			// prioritize slots over sizes
-			if sizeIndex == 0 {
-				prioritizedRequests = append(prioritizedRequests, request)
-			} else {
-				requests = append(requests, request)
-			}
-		}
-	}
-
-	// cap the number of requests to requestLimit
-	requests = append(prioritizedRequests, requests...)
-	if len(requests) > requestLimit {
-		requests = requests[:requestLimit]
-	}
-
-	if len(requests) == 0 {
-		return nil, &errortypes.BadInput{
-			Message: "Invalid ad unit/imp/size",
+	imps := request.Imp
+	for iImp := range imps {
+		request.Imp = imps[iImp : iImp+1]
+		if request.Site != nil {
+			setSitePublisherId(request, iImp)
 		}
-	}
 
-	ch := make(chan ixBidResult)
-	for _, request := range requests {
-		go func(bidder *pbs.PBSBidder, request callOneObject) {
-			result, err := a.callOne(ctx, request.requestJSON)
-			result.Request = &request
-			result.Error = err
-			if result.Bid != nil {
-				result.Bid.BidderCode = bidder.BidderCode
-				result.Bid.BidID = bidder.LookupBidID(result.Bid.AdUnitCode)
-				result.Bid.Width = request.width
-				result.Bid.Height = request.height
-
-				if result.Bid.BidID == "" {
-					result.Error = &errortypes.BadServerResponse{
-						Message: fmt.Sprintf("Unknown ad unit code '%s'", result.Bid.AdUnitCode),
+		if request.Imp[0].Banner != nil {
+			banner := *request.Imp[0].Banner
+			request.Imp[0].Banner = &banner
+			formats := getBannerFormats(&banner)
+			for iFmt := range formats {
+				banner.Format = formats[iFmt : iFmt+1]
+				banner.W = openrtb.Uint64Ptr(banner.Format[0].W)
+				banner.H = openrtb.Uint64Ptr(banner.Format[0].H)
+				if requestData, err := createRequestData(a, request, &headers); err == nil {
```

Comment:
It doesn't look like `createRequestData` errors are sent back to the publisher. Please consider adding the err to the errs slice.

Reply:
`json.Marshal(*openrtb.BidRequest)` can never fail:
https://github.com/golang/go/blob/master/src/encoding/json/encode.go

 That's why everyone has a problem with the unit test coverage of the "failed" branch.

Reply:
I don't see where in the documentation it says `json.Marshal` can never fail.

It can fail for a few reasons, such as if the struct contains unmarshallable types (chan, func pointer, etc...), if there are cyclical references, or if there is memory corruption. The latter is what I'm more concerned with happening as we've seen it before.

We understand it is not feasible to reproduce a `json.Marshal` error for tests and we don't comment on lack of coverage for those parts.

Reply:
What I meant is not that `json.Marshal` can never fail (it can fail for the reasons you've mentioned), but that `json.Marshal(*openrtb.BidRequest)` can never fail because it doesn't have these cases. If we have a memory corruption it's a much bigger problem that can't be solved with an err flag.